p2p/messenger.go

```diff
@@ -0,0 +1,342 @@
+	ggio "github.com/gogo/protobuf/io"
+	"github.com/jbenet/go-context/io"
+	inet "github.com/libp2p/go-libp2p-net"
+	"github.com/libp2p/go-libp2p-peer"
+)
+
+// ErrReadTimeout must be used when receiving timeout while reading
+const ErrReadTimeout = errors.Error("timed out reading response")
+
+// ErrInvalidatedMessageSender must be used when the message sender object created is no longer valid (connection has dropped)
+const ErrInvalidatedMessageSender = errors.Error("message sender has been invalidated")
+
+type bufferedWriteCloser interface {
+	ggio.WriteCloser
+	Flush() error
+}
+
+// The Protobuf writer performs multiple small writes when writing a message.
+// We need to buffer those writes, to make sure that we're not sending a new
+// packet for every single write.
+type bufferedDelimitedWriter struct {
+	*bufio.Writer
+	ggio.WriteCloser
+}
+
+func newBufferedDelimitedWriter(str io.Writer) bufferedWriteCloser {
+	w := bufio.NewWriter(str)
+	return &bufferedDelimitedWriter{
+		Writer:      w,
+		WriteCloser: ggio.NewDelimitedWriter(w),
+	}
+}
+
+func (w *bufferedDelimitedWriter) Flush() error {
+	return w.Writer.Flush()
+}
+
+type p2pMessenger struct {
+	host host.Host // the network services we need
+	self peer.ID   // Local peer (yourself)
+
+	timout time.Duration
+	ctx    context.Context
+
+	strmap map[peer.ID]map[protocol.ID]*messageSender
+	smlk   sync.Mutex
+
+	plk sync.Mutex
+
+	msgHandlers map[pb.MessageType]func(ctx context.Context, peer peer.ID, protoc protocol.ID, msg *pb.P2PEnvelope) (*pb.P2PEnvelope, error)
+}
+
+func newP2PMessenger(ctx context.Context, host host.Host, p2pTimeout time.Duration) *p2pMessenger {
+	return &p2pMessenger{
+		ctx:         ctx,
+		host:        host,
+		self:        host.ID(),
+		timout:      p2pTimeout,
+		strmap:      make(map[peer.ID]map[protocol.ID]*messageSender),
+		msgHandlers: make(map[pb.MessageType]func(ctx context.Context, peer peer.ID, protoc protocol.ID, msg *pb.P2PEnvelope) (*pb.P2PEnvelope, error))}
+}
+
+// init initiates listening to given set of protocol streams
+func (mes *p2pMessenger) init(protocols ...protocol.ID) {
+	for _, p := range protocols {
+		mes.host.SetStreamHandler(p, mes.handleNewStream)
+	}
+}
+
+// addHandler adds a message handler for a specific message type
+func (mes *p2pMessenger) addHandler(mType pb.MessageType, handler func(ctx context.Context, peer peer.ID, protoc protocol.ID, msg *pb.P2PEnvelope) (*pb.P2PEnvelope, error)) {
+	mes.msgHandlers[mType] = handler
+}
+
+// handleNewStream implements the inet.StreamHandler
+func (mes *p2pMessenger) handleNewStream(s inet.Stream) {
+	go mes.handleNewMessage(s)
+}
+
+func (mes *p2pMessenger) handleNewMessage(s inet.Stream) {
+	ctx := mes.ctx
+	// ok to use. we defer close stream in this func
+	cr := ctxio.NewReader(ctx, s)
+	cw := ctxio.NewWriter(ctx, s)
+
+	// delimited readers and writers to set length of the protobuf messages to the stream
+	r := ggio.NewDelimitedReader(cr, inet.MessageSizeMax)
```

Comment:
This value `MessageSizeMax` I assume is the batch reading size and does not mean that any part of a proto object bigger than that will be discarded otherwise it is a problem.

Reply:
I also thought about this, it does discard right now, but we can increase it arbitrarily or even remove it by forking ggio reader and writer, so it should be fine.